### DISCUSSION TITLE
Backref cascade delete works on session.delete but not with session.execute(delete(...))

### DISCUSSION THREAD

**Discussion by lafrech** [sqlalchemy/sqlalchemy#10672] [Usage Questions] [answered]
I'm using cascade delete as explained in [Cascades](https://docs.sqlalchemy.org/en/20/orm/cascades.html).

```py
class Order(Base):
    __tablename__ = "order"

    items = relationship("Item", cascade="all, delete-orphan")
    customer = relationship("User", cascade="save-update")


class Item(Base):
    __tablename__ = "item"

    order_id = sqla.Column(
        Integer,
        sqla.ForeignKey("order.id"),
        nullable=False,
    )
    order = relationship(
        "Order", backref=backref("items", cascade="all, delete-orphan")
    )
```

It all words fine when I delete an element like this:

```py
session.delete(item)
```

But it doesn't cascade delete (and therefore raises an error) when the deletion is done like this:

```py
stmt = (
    sqla.delete(Order)
    .where(whatever)
)
session.execute(stmt)
```

Is this expected behaviour?

It might be to someone more aware of the concepts in SQLAlchemy but it came as a surprise to me. Most of the time, I develop a CRUD API, where all deletes are of the first form in a "delete by ID" resource, so I never got caught by it until today.

It works fine (and is probably more efficient) if I use the database ON DELETE feature (PostgreSQL inside).

```py
class Item(Base):
    __tablename__ = "item"

    order_id = sqla.Column(
        Integer,
        sqla.ForeignKey("order.id", ondelete="CASCADE"),
        nullable=False,
    )
    order = relationship(
        "Order", backref=backref("items", cascade="all, delete-orphan", passive_deletes=True)
    )
```

(BTW, `backref` is documented as [legacy](https://docs.sqlalchemy.org/en/20/orm/backref.html) and users are told to use `back_populates` but the Cascade does still use `backref`.)

**Comment by zzzeek** [comment, marked as answer]
hi -

I have to apologize as it looks like in our rewrite of the docs for 2.0, where we would assume users learn about update() and delete() with the ORM [here](https://docs.sqlalchemy.org/en/20/orm/queryguide/dml.html#orm-update-and-delete-with-custom-where-criteria), we lost all of our important notes regarding this.  these notes can be seen in the 1.4 docs at https://docs.sqlalchemy.org/en/14/orm/session_basics.html#orm-expression-update-delete in the red box.

TLDR,  session.execute(delete()) just emits a DELETE statement. cascade effects are not in play because this method does not assume any of the affected objects are actually present.

will restore those notes to the new docs

**Comment by lafrech** [comment]
Thanks for answering.

> TLDR, session.execute(delete()) just emits a DELETE statement. cascade effects are not in play because this method does not assume any of the affected objects are actually present.

I'm not sure I get the rationale. What I expected, I guess, without really thinking about the implications, was that the ORM would automagically select and delete all children before the parent deletion (unless ON DELETE + passive deletes). While it does work for single deletes, it does not for session.execute but I don't understand in your answer if it doesn't because that would be too complicated to implement or because it is not something we want (in which case I don't understand why we wouldn't want it). If it is a perf issue, it could be an opt-in.

**Reply by zzzeek** [reply]
it would be massively complicated and not even possible because cascades can go across any number of levels deep
